woocommerce/woocommerce-gutenberg-products-block
# Cart and checkout should respect the global "Hide shipping costs until an address is entered" setting

This PR fixes shipping discrepancies during cart and checkout caused by the global `woocommerce_shipping_cost_requires_address` setting. When `woocommerce_shipping_cost_requires_address ` is true, shipping total from the cart is **always** 0. This conflicts with the block level setting `Hide shipping costs until an address is entered` which has no control over this.

1. Removes the block level setting "Hide shipping costs until an address is entered" since there is no way to make the cart totals respect this and ignore the global setting.
2. The REST API will return `null` shipping totals if calculations have not been performed by core.
3. The REST API will return an empty array of shipping rates if calculations have not been performed by core.
4. Updated placeholders when shipping has not been returned in the above circumstances.
5. Fixed an issue where all fields were optional in the shipping calculator, since those fields are required by core on order to return rates.

Fixes #3148

### Screenshots

If no rates are returned from the API (customer has no input their full shipping address), the shipping section will show a message instead of the list of current rates:

<img width="557" alt="Checkout – one wordpress test 2020-11-09 12-45-02" src="https://user-images.githubusercontent.com/90977/98544050-4fd3a100-228b-11eb-8ea1-d8d20f70933f.png">

And in the summary area a similar message is shown:

<img width="306" alt="Checkout – one wordpress test 2020-11-09 12-44-33" src="https://user-images.githubusercontent.com/90977/98544075-58c47280-228b-11eb-9ff4-ac9a45c7177c.png">

This is instead of showing a totals row with 0 cost which is not accurate.

### How to test the changes in this Pull Request:

1. Before running this PR, in Settings > Shipping > Shipping Options, turn on the option to require an address
2. Setup a single shipping rate
3. Edit the cart block and disable "Hide shipping costs until an address is entered"
4. In incognito, go to cart and see that the shipping rate is selected, but the shipping cost is 0. This is the issue reported in #3148.

Then apply this PR and confirm:

1. The messaging in cart and checkout states that shipping will be shown after providing an address.
2. After entering your address, rates are then displayed and the total cost is not 0
3. The "Hide shipping costs until an address is entered" block level option is no longer displayed when editing the block.

This is pending feedback from @nerrad since it removes the block level option and needs design approval.

### Changelog

> Ensures cart and checkout respect the global "Hide shipping costs until an address is entered" setting.
